File: NeetCode/BinarySearch/Medium/33.SearchRotatedArray.py
# Given the array nums after the possible rotation and an integer target, 
# return the index of target if it is in nums, or -1 if it is not in nums.

def search(nums, target):
    left, right = 0, len(nums) - 1

    while left <= right:
        mid = (right + left)//2

        if nums[mid] == target:
            return mid

        if nums[mid] >= nums[left]:
            if nums[mid] < target or target < nums[left]:
                left = mid + 1
            else:
                right = mid - 1
        else:
            if target < nums[mid] or target > nums[right]:
                right = mid - 1
            else:
                left = mid + 1

    return -1

print(search([4,5,6,7,0,1,2], 0))
print(search([4,5,6,7,0,1,2], 3))
print(search([1], 0))

File: swift/Neetcode/BinarySearch/Medium/33.SearchRotatedArray.swift
// Given the array nums after the possible rotation and an integer target, 
// return the index of target if it is in nums, or -1 if it is not in nums.

func search(_ nums: [Int], _ target: Int) -> Int {
    var left = 0, right = nums.count - 1

    while left < right {
        let mid = (right + left) / 2

        if nums[mid] == target { return mid }

        if nums[left] <= nums[mid] {
            if target >= nums[left], target < nums[mid] {
                right = mid
            } else {
                left = mid + 1
            }
        } else {
            if target > nums[mid], target <= nums[right] {
                left = mid + 1
            } else {
                right = mid
            }
        }
    }

    return nums[left] == target ? left : -1
}

print(search([4,5,6,7,0,1,2], 0))
print(search([4,5,6,7,0,1,2], 3))
print(search([1], 0))